Sources/Fuzzilli/Core/ProgramBuilder.swift

```diff
@@ -302,23 +302,31 @@ public class ProgramBuilder {
         return randVar(ofType: type)
     }
 
+    /// Returns a random variable from the outer scope.
+    public func randVarFromOuterScope() -> Variable {
+        assert(hasVisibleVariables)
+        return randVarInternal(fromOuterScope: true)!
+    }
+
     /// Returns a random variable satisfying the given constraints or nil if none is found.
-    private func randVarInternal(_ selector: ((Variable) -> Bool)? = nil) -> Variable? {
+    private func randVarInternal(_ selector: ((Variable) -> Bool)? = nil, fromOuterScope: Bool = false) -> Variable? {
```

Comment:
Since this variable selection logic is pretty sensitive and bugs in it wouldn't necessarily be detected, could you add a few short tests for this that just do something like:

- Open 2-3 scopes, have a single variable in the innermost scope (the outer ones are empty), assert that `randVar` gives you that one (and maybe that `randVarInternal(fromOuterScope: true)` returns nil if that's easy to do)
- Open e.g. 3 scopes, have at least one variable in the innermost scope and a single variable in the 2nd scope, assert that `randVarFromOuterScope` gives you that one
- Open 2-3 scopes, have a couple variables in each, assert that `randVarInternal({ $0 == X })` gives you the variable X for some variable X from each scope. You might need to remove `private` from this method for that to work, but I think that'd be fine.
- Same as above, but with `randVarInternal(fromOuterScope: true, { $0 == X })`

That'd be great, thanks!